library/common/extensions/filter/c_types.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include "library/common/types/c_types.h"
+
+// NOLINT(namespace-envoy)
+
+/**
+ * Return codes for header filter invocations. See envoy/http/filter.h
+ */
+typedef enum {
+  ENVOY_FILTER_HEADERS_STATUS_CONTINUE,
+  ENVOY_FILTER_HEADERS_STATUS_STOP_ITERATION,
+  ENVOY_FILTER_HEADERS_STATUS_CONTINUE_AND_END_STREAM,
+  ENVOY_FILTER_HEADERS_STATUS_STOP_ALL_ITERATION_AND_BUFFER,
+} envoy_filter_headers_status_t;
+
+/**
+ * Return codes for data filter invocations. See envoy/http/filter.h
+ */
+typedef enum {
+  ENVOY_FILTER_DATA_STATUS_CONTINUE,
+  ENVOY_FILTER_DATA_STATUS_STOP_ITERATION_AND_BUFFER,
+  ENVOY_FILTER_DATA_STATUS_STOP_ITERATION_NO_BUFFER,
+} envoy_filter_data_status_t;
+
+/**
+ * Function signature for header filter invocations.
+ */
+typedef envoy_filter_headers_status_t (*envoy_filter_on_headers_f)(envoy_headers headers,
+                                                                   bool end_stream, void* context);
+
+/**
+ * Function signature for data filter invocations.
+ */
+typedef envoy_filter_data_status_t (*envoy_filter_on_data_f)(envoy_data data, bool end_stream,
+                                                             void* context);
+
+/**
+ * Raw datatype containing dispatch functions for a platform-native HTTP filter. Leveraged by the
+ * PlatformHarnessFilter
+ */
+typedef struct {
+  envoy_filter_on_headers_f on_request_headers;
+  envoy_filter_on_data_f on_request_data;
+  envoy_filter_on_headers_f on_response_headers;
+  envoy_filter_on_data_f on_response_data;
+  void* context;
+} envoy_http_filter;
```

Comment:
Do we want to create a distinction between a request(decode) and response(encode) filter? And a sub-type that is able to do both? What would be a good pattern to do that in c?

Reply:
I mean you could use structs that have identical layouts but reserve the space for the missing callbacks. Or you could just use this one struct and set callbacks you're not using to null. I think the latter simplifies things.

Reply:
But then you would have to do null checks or assertions, vs. having the type dictate what functions you should implement.

Reply:
My plan was to default all callbacks to null and handle that gracefully in the base (bridge) implementation. I think it's more defensive, given the type variation.